lib/workers/repository/package-files.ts

```diff
@@ -24,7 +26,104 @@ export class PackageFiles {
     this.data.clear();
   }
 
-  public static getDashboardMarkdown(config: RenovateConfig): string {
+  /**
+   * Removes the last dependency/entry in the PackageFiles data
+   * i.e. the last line in the tobe generated detected dependency section
+   * @Returns true if anything that translates to a markdown written line was deleted
+   *          otherwise false is returned
+   */
+  private static pop(): boolean {
+    // get detected managers list of the last listed base branch
+    const [branch, managers] = Array.from(this.data).pop() ?? [];
+    if (!branch) {
+      return false;
+    }
+
+    // delete base branch listing if it has no managers left
+    if (!managers || is.emptyObject(managers)) {
+      return this.data.delete(branch);
+    }
+
+    // get all manifest files for the last listed manager
+    const [manager, packageFiles] = Object.entries(managers).pop() ?? [];
+
+    // delete current manager if it has no manifest files left
+    if (!packageFiles || is.emptyArray(packageFiles)) {
+      return delete managers[manager!];
+    }
+
+    // delete manifest file if it has no deps left
+    const len = packageFiles.length - 1;
+    if (is.emptyArray(packageFiles[len].deps)) {
+      return !!packageFiles.pop();
+    }
+
+    // remove the last listed dependency
+    return !!packageFiles[len].deps.pop();
+  }
+
+  /**
+   * Truncates the detected dependencies' section until it fits the available space
+   * i.e. It has length smaller than maxLength.
+   * This does not mutate the original PackageFiles data
+   * Note:  setHeader=false is used for testing purposes only
+   *        Mainly for comparing the output of getTruncatedMarkdown and getDashboardMarkdown
+   * @param config
+   * @param maxLength
+   * @param setHeader
+   */
+  public static getTruncatedMarkdown(
+    config: RenovateConfig,
+    maxLength: number,
+    setHeader = true
+  ): string {
+    const org = this.data; // backup data
+    // deep clone data
+    this.data = new Map(clone(Array.from(this.data))); // only mutate cloned data
```

Comment:
why to we need to clone?

Reply:
we should probably change `this.data` to `Record<string, Record<string, PackageFile[]> | null>` for easier handling?

Reply:
it is cloned as the original data is mutated, and i didnt want this to have side effects on the data.
i dont know if this data is being used in some other part.

Reply:
it should then be cloned here

https://github.com/renovatebot/renovate/blob/e1a2930f254ee67446c8ddfbd0c8f2c7d5f0b571/lib/workers/repository/process/fetch.ts#L130

or here:
https://github.com/renovatebot/renovate/blob/e1a2930f254ee67446c8ddfbd0c8f2c7d5f0b571/lib/workers/repository/package-files.ts#L16

so we don't need to clone again

Reply:
> we should probably change this.data to Record<string, Record<string, PackageFile[]> | null> for easier handling?

i think this is better be done in a separate PR.

![image](https://user-images.githubusercontent.com/97394622/182323517-df7460ea-b558-4c79-9514-582abdcadc9c.png)

Reply:
> it should then be cloned here

This will only clone the package files object, but not the keys.
so if we delete a base branch(key) its data will be lost although its package file is cloned.
but i think it should be cloned regardless.


having said that, i still think the cloning in `getTruncatedMarkdown` is needed as we might want to use this data in the future (?).

what about only cloning if truncation is needed?

Reply:
now cloning when setting the packageFile data.
and also backing up the data only if truncation is needed.
restoring afterwards.